### PR TITLE
FSE: Fall back to next best template in hierarchy when querying through REST API (Take Two)

### LINKED ISSUE #36648: Block Template creation is using empty content instead of the appropiate fallback
### Description

This was brought to my attention by @carolinan over at https://github.com/WordPress/twentytwentytwo/issues/256

When a theme doesn't provide a template for a specific page (404, Search for example) we will always fallback through the hierarchy and show in the frontend an alternative existing template. In the case of TT2, if the search.html doesn't exist, the frontend will show index.html instead:

Screenshot 2021-11-19 at 13 49 46

This is inconsistent with what the flow of template creation is showing us. When you try to create a new Search template, we see no content at all:

Screenshot 2021-11-19 at 13 49 24

Screenshot 2021-11-19 at 13 49 40

This is inconsistent with what the user sees in the frontend before creating the template and forces them to create something from scratch.

### Step-by-step reproduction instructions

1. Install TT2, check in the frontend what a search results page returns
2. Go to the Site Editor and try to create a Search template
3. The Search template is empty.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ockham**
## What?

Take Two of #37258.

Purports to fix #36648, as an alternative approach to #37054. For more background, see https://github.com/WordPress/gutenberg/pull/37054#issuecomment-988271255

_(More description to come)_

## Why?
See e.g. https://github.com/WordPress/gutenberg/pull/37258#issuecomment-1136144085 and https://github.com/WordPress/gutenberg/pull/37258#issuecomment-1161697393.

## How?

By always using the full template hierarchy to look for potential fallback templates, even when queried for a template ID.

## Testing Instructions

This testing routine will verify a number of things:
- That if a new template is created, it will pre-populate the editor with the "next best", less specific, fallback template.
  - This can be either a manually created template, or a theme-supplied one.
- That the template titles and descriptions are correct.

First, we create a "Single Post" template and verify that it comes pre-populated with the contents of the theme-supplied "Single" template:

1. Enable the Twenty Twenty-Two theme.
2. Go to the Site Editor.
3. Click the "W" logo menu in the top left corner.
4. In the sidebar that opens, select "Templates".
6. Click the "Add New" button in the top right corner.
7. In the dropdown that opens, click on the "Single item: Post" template.
8. In the modal that opens, click on "All Posts -- For all items".
11. Verify that the site editor opens, pre-populated with the generic "single" template. (This is to ensure parity with the frontend, which would also use the "single" template as a fallback when there's no dedicated single-_post_ template.)
12. Verify that the title and description are still as prior to this PR ("Single item: Post" --  "Displays a single item: Post.")
13. Make a change to the template. Make sure it's not to the header or footer template parts, nor to the post content. Ideally, insert a Paragraph block with some text right above the footer (and below the content). (Verify that it's a direct child of the template by checking in the bottom breadcrumbs bar that it shows up as "Template > Paragraph".)
9. Save the template.
10. Once saved, navigate back to the "Templates" screen (via the top left "W" button that opens the sidebar).
14. Verify that there's now a "Single item: Post" template at the top of the list of the theme's templates.
15. Verify that it's distinct from the "Single" template at the bottom of the list.

Second, we create a "Single Post" template for a specific post and verify that it comes pre-populated with the generic "Single Post" template that we just created: **This is still buggy, reload every time you go back to the template list**

1. In the template list, click the "Add New" button in the top right corner.
2. Since you've already created the generic "Single item: Post" template, the editor will directly open a modal to ask you for _which_ post you'd like to create a template. 
10. Select a post from that list, e.g. the standard WordPress "Hello World" post.
11. Verify that the site editor opens, with the "Single item: Post" template contents loaded that you just created. (This is again to reflect parity with the frontend).
12. Verify that the template content has the change that you just made earlier to the "Single item: Post" template.
13. Verify that the title and description are still as prior to this PR ("Post: Hello world!" -- "Template for Post: Hello world!").
14. Make another change to the template, ideally to the part that you changed to distinguish the "Single item: Post" template.
9. Save the template.
10. Once saved, navigate back to the "Templates" screen (via the top left 'W' button that opens the sidebar).
15. Verify that there's now a "Post: Hello World!" template at the top of the list of the theme's templates.
16. Verify that it's distinct from the "Single item: Post" template right below.

Ideally, give it some more smoke testing -- whatever you can think of makes sense: delete the template again; find a way to test it on the frontend; etc.

## TODO

- [x] Use the proper template title (i.e. "Single Post: Hello World" instead of "Single") and description.
- [x] Fix issue with blank template when creating second template in a row ([see](https://github.com/WordPress/gutenberg/pull/41848#issuecomment-1167477225)).
- [ ] Fix issue with changes to post-specific templates (e.g. `single-post-hello-world`) being applied to existing CPT-specific templates (e.g. `single-post`).
- [ ] Integrate the Home/Frontpage fallback logic (`gutenberg_resolve_home_template`/`_resolve_home_block_template`).
- [ ] We might be able to remove some client-side logic, or infer it via REST API (if we also expose some template titles).

## Screenshots
TBD

### CODE REVIEW

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 1]
```diff
@@ -195,34 +211,58 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
...
+					// translators: %s: Name of the post type e.g: "Post".
+					__( 'Displays a single item: %s.' ),
+					$labels->singular_name
+				);
+			}
 		}
+	} elseif ( array_key_exists( $slug, $default_template_types ) ) {
+		$block_template->title       = $default_template_types[ $slug ]['title'];
+		$block_template->description = $default_template_types[ $slug ]['description'];
 	}
```
Comment (ockham) [inline lib/compat/wordpress-6.1/block-template-utils.php]: Need to encapsulate this logic in a function.

**Comment by ockham** [comment]
👋 FYI @jameskoster @ntsekouras 🙂

**Comment by jameskoster** [comment]
Woohoo! I'm very excited about this one. 

The mechanics seem to work very well – the correct fallback template was used in all of my testing. 

I did notice however that new templates are not immediately published upon creation (I believe they are on trunk), but it's also not possible to save when you land in the editor. This can lead to some confusing UX:

https://user-images.githubusercontent.com/846565/175540977-1830dd80-210a-4ac7-be24-4c94507160ef.mp4

I'm not sure that auto-publishing is the correct approach, but that's probably something to explore separately. For now I suppose it's more important to be consistent.

---

I performed a number of additional tests by deleting all templates except Index in my current theme.

Not wanting to add more complexity (😬), but we might consider some exceptions, or even some hard-coded fallbacks. A couple of situations felt a bit odd:

1. Creating a 404 template based on Index doesn't make all that much sense
2. Creating a singular template based on Index also doesn't feel right

We can probably explore these in a follow-up since they're flows more likely to be encountered by advanced users, where the issues aren't so egregious.

---

An odd quirk I observed: During the first creation, the fallback works as expected. But on subsequent creations of different templates I am met with a blank canvas, unless I return to the templates list and repeat the action. Better explained with a video:

https://user-images.githubusercontent.com/846565/175541841-c7795862-5786-44e2-b855-2693c6b4e173.mp4

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 2]
```diff
@@ -195,34 +211,58 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
...
+		if ( 'single' === $slug_parts [0] ) {
+			// Get CPT labels
+			$post_type = get_post_type_object( $slug_parts[1] );
+			$labels    = $post_type->labels;
 
-		if ( ! is_wp_error( $template ) ) {
-			return $template;
+			if ( count( $slug_parts ) > 2 ) {
+				// Now we look for the CPT with slug as defined in $slug_parts[2]
+				$post                  = get_page_by_path( $slug_parts[2], OBJECT, $slug_parts[1] );
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/block-template-utils.php]: Why use this function over a WP_Query?

Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/block-template-utils.php]: Also can't we just do this? 
```suggestion
				$post                  = get_page_by_path( $slug_parts[2], OBJECT, $template_type );
```

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
This PR is pretty hard to follow and the problem it is trying to solve seems unclear.

In short, I don't like this solution at all. `get_block_template` is now a core function and is being used by developers. Out of the blue changing the behaviour of it, without warning and with no test coverage, feel reckless. Adding magic logic to this function to fix this problem, is not the place to do this. A solution that fallbacks back in javascript makes much more sense to me. 

I would recommend adding a link to the REST API response or a default content field maybe.

**Comment by ockham** [comment]
Thank you for your candid feedback @spacedmonkey. I realize this approach is not without controversy — especially the part about using the template fallback even when requesting a template by a given ID (see prior discussion at #37258).

> This PR is pretty hard to follow and the problem it is trying to solve seems unclear.

Well, it’s still a work in progress ;) FWIW, I plan to encapsulate some logic into functions to better organize the code.
OTOH, there is some complexity that’s inherent in WordPress’ template hierarchy resolution, and I believe there’s barely any prior art that attempts to fully emulate it in the editor, so some initial confusion might be inevitable. I plan to mitigate this by documenting some design decisions and how the "emulation" works, either in the PR description, or in some GB docs.

> In short, I don't like this solution at all. `get_block_template` is now a core function and is being used by developers. Out of the blue changing the behaviour of it, without warning and with no test coverage, feel reckless. Adding magic logic to this function to fix this problem, is not the place to do this.

FWIW, I definitely want to add some test coverage. That said, changing `get_block_template` is likely the most controversial part. However, I wouldn’t be filing this PR if I wasn’t convinced that it was logically sound enough to at least stand the chance of proving itself by a good amount of testing. Furthermore, I did an [audit](https://github.com/WordPress/gutenberg/pull/37258#issuecomment-990895049) of all call sites of `get_block_template` when filing this PR’s predecessor. I’m happy to revisit to cover any new or changed call sites.

> A solution that fallbacks back in javascript makes much more sense to me.

I respectfully disagree here. What we currently have in Gutenberg (and WordPress’s REST API) is an incomplete emulation of the template hierarchy resolution as it works on the frontend that is, among other things, ignorant of per-ID and per-slug templates (such as `single-post-my-post` or `page-123`), or of fallbacks across template types (e.g. frontpage to home).
While [logic has been added to cover the latter example](https://github.com/WordPress/wordpress-develop/blob/ae9cff730f54f2401ce3271477b1ba63fed81f83/src/wp-includes/block-template.php#L335-L365), it still doesn’t cover the entirety of the template hierarchy; plus it’s done in a bit of an ad-hoc way, via a filter.

IMO, adding JavaScript to the mix to fully capture the template hierarchy makes things worse: Developers would then have to look into both PHP and JS files to figure out how template lookup works; and since the delineation would be somewhat arbitrary, it’d be fairly non-obvious how to extend that logic.

I'll continue to work a bit more on this PR and to provide more background and information how it's supposed to work etc. I'd be grateful if you could give it another look once it's ready for review! :)

**Comment by ockham** [comment]
> Woohoo! I'm very excited about this one.
>
> The mechanics seem to work very well – the correct fallback template was used in all of my testing.

Hey @jameskoster! Happy to hear, and thank you for testing! 🙌 

> I did notice however that new templates are not immediately published upon creation (I believe they are on trunk), but it's also not possible to save when you land in the editor. This can lead to some confusing UX: [video]

Yeah, so this behavior mostly results from the editor not enabling the Save button if the user hasn't made any changes to the previously "saved" state of the template -- and in this case, the notion of "saved" is extended to what the fallback template looked like. So the Save button should become enabled once you make any change to the template.

OTOH, if you don't make any changes and instead return straight to the template list, the template isn't saved indeed -- for the same reason: WordPress will continue to use the fallback template to render whatever the more specific template would've been for.

So, for example:
- You click "Add new" to add a single post template for the "Hello World" blogpost.
- The Site Editor opens, pre-populated with the "generic" single post template.
- You decide that you don't actually want to make any changes and go back to the template list.
- No new template shows up. Instead, WordPress will continue to use the "generic" single post template to render the "Hello World" blogpost on the frontend.

I think that this might be fairly intuitive in some cases (like the above one maybe), but since this behavior applies to _all_ templates, it's arguable that it's less so for more "exotic" fallbacks.

LMK if this makes sense. We can try to change some of this behavior; the complexity involved may vary greatly depending on what aspect(s) we'd like to change specifically, since some things tie pretty deeply into existing concepts and might be hard to change without big changes to the underlying functionality. So it'd be good to get a sense of what things are a dealbreaker UX-wise, and which ones we can maybe mitigate through messaging etc :)

**Comment by ockham** [comment]
> I performed a number of additional tests by deleting all templates except Index in my current theme.
>
> Not wanting to add more complexity (😬), but we might consider some exceptions, or even some hard-coded fallbacks. > A couple of situations felt a bit odd:
>
> 1. Creating a 404 template based on Index doesn't make all that much sense
> 2. Creating a singular template based on Index also doesn't feel right

The thing is that this PR is seeking the [template hierarchy](https://developer.wordpress.org/themes/basics/template-hierarchy/) as it would work on the frontend, and those are the actual fallbacks that would be used there 😕 (also see https://wphierarchy.com/)

Now I agree that those don't _feel_ right; they probably haven't been much of a problem prior to FSE since themes would pretty much always include a 404 and single post template, so the fallback was rarely needed. While I'd love to make the fallbacks fell more right, I also don't want to deviate from how it'd work on the frontend. Not totally sure how to best tackle this 🤔 

> We can probably explore these in a follow-up since they're flows more likely to be encountered by advanced users, where the issues aren't so egregious.

Sounds good; definitely something we can work on later as a refinement on top of this PR 👍

**Comment by ockham** [comment]
> An odd quirk I observed: During the first creation, the fallback works as expected. But on subsequent creations of different templates I am met with a blank canvas, unless I return to the templates list and repeat the action. Better explained with a video: [video]

Yeah, I've also run into this behavior. That's definitely a bug -- probably some sort of editor state that's kept around. I'll look into this.

File: lib/compat/wordpress-6.1/block-template-utils.php (hunk 2 above)
Comment (ockham) [reply lib/compat/wordpress-6.1/block-template-utils.php]: `WP_Query` is probably the better option; `get_page_by_path` was the stop-gap to make it work 👍

Comment (ockham) [reply lib/compat/wordpress-6.1/block-template-utils.php]: I'm afraid that won't work -- `$template_type` will be something like `single`, `category`, or even `404`.

If it's `single`, then `$slug_parts[1]` will be the _post_ type that the template is for.

(`$template_type` OTOH will be reflected in `$slug_parts[0]`.)

**Comment by ockham** [comment]
> > An odd quirk I observed: During the first creation, the fallback works as expected. But on subsequent creations of different templates I am met with a blank canvas, unless I return to the templates list and repeat the action. Better explained with a video: [video]
> 
> Yeah, I've also run into this behavior. That's definitely a bug -- probably some sort of editor state that's kept around. I'll look into this.

Should be fixed now.

There's still an error where when creating a CPT-specific template (Single item: Post) followed by a post-specific template (Post: Hello World), the latter will somehow override the former. I'll investigate tomorrow.

File: lib/compat/wordpress-6.1/block-template-utils.php (hunk 1 above)
Comment (ockham) [reply lib/compat/wordpress-6.1/block-template-utils.php]: (Done in d1ee1f7a09ba572ef2ea299826c4d81b3696f59a)

**Comment by jameskoster** [comment]
> LMK if this makes sense

It makes sense, but the UX is confusing because it is directly opposed to what happens when you create a template on trunk. There the template is published immediately, so for that not to happen in this PR feels broken to me. 

At the moment it is probably best to align with trunk, IE publish immediately. We can then explore whether that is the best flow separately. 

> Not totally sure how to best tackle this

The fallback is least successful in two key flows, creating a `404` template, and creating any singular template (`page`, `single`, etc) where `singular` doesn't already exist. So it may be that we need to have a static fallback template for these two scenarios.

**Comment by ockham** [comment]
I've now started an alternative exploration over in #42007, based on feedback here and via DM. Its overall footprint is smaller, and templates are saved immediately after creating 🙂

**Comment by paaljoachim** [comment]
Testing using the Gutenberg PR build.
Using Twenty Twenty Two.
WP 6.0.1

Part 1.

8.
> Verify that the site editor opens, pre-populated with the generic "single" template. (This is to ensure parity with the frontend, which would also use the "single" template as a fallback when there's no dedicated single-post template.)

9.
> Verify that the title and description are still as prior to this PR ("Single item: Post" -- "Displays a single item: Post.")

I can verify that the site editor opens, pre-populated with the generic "single" template, and that I see Single item: Post

13.
> Verify that there's now a "Single item: Post" template at the top of the list of the theme's templates.

14.
> Verify that it's distinct from the "Single" template at the bottom of the list.

This is what I see (The site is a local testing site I have been using for all things FSE).

Screenshot 2022-07-17 at 15 19 29

------------

Part 2.

4.
> Verify that the site editor opens, with the "Single item: Post" template contents loaded that you just created. (This is again to reflect parity with the frontend).

5.
> Verify that the template content has the change that you just made earlier to the "Single item: Post" template.

6.
> Verify that the title and description are still as prior to this PR ("Post: Hello world!" -- "Template for Post: Hello world!").

I can verify 4, 5, and 6.

10.
> Verify that there's now a "Post: Hello World!" template at the top of the list of the theme's templates.

NB!
**I am not seeing a "Post: Hello World" in the Template list.**

I again click the Add New button -> Single item: Post and the Add template: Post modal is seen.
Behind and below the modal is seen a Single Post made by me.

Screenshot 2022-07-17 at 15 33 45

Upon making another Template for all single posts I see this:
Screenshot 2022-07-17 at 15 36 09

So there is some weirdness going on.

I am starting over. I have now deleted all my earlier made custom templates.
I am now seeing at the top: 
Single item: Post
Displays a single item: Post.

I made a template for Hello World but again it did not show up in the Templates List screen.

**Comment by ockham** [comment]
Thank you very much for your testing, @paaljoachim!

Unfortunately, I'm afraid we're not going to move forward with this PR since it's proven too controversial. I've filed a somewhat smaller change in #42007, but even that one looks like it's not everyone's preferred solution. We're probably going to go with @ntsekouras' #42520 😊 

Anyway, I'm going to close this PR. Sorry -- should've done it sooner!

**Comment by paaljoachim** [comment]
Thanks for working on it @ockham Bernie!
